Test officially supported cmd taps.

MikeMcQuaid opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

All of these taps use Homebrew internal APIs (or will shortly) and we autoinstall them all from `brew $CMD`. We should adjust our CI to ensure that we never accidentally break these taps when making changes to core code so that these taps can rely more on this core code rather than having to e.g. vendor equivalent code that never changes on our end.

CI may well by wrong here as I've not tested this on the actual bot and it won't run `brew tests --official-cmd-taps` (although I should fix that so we can test new `brew test-bot` changes here).

The approach may well be terrible; we could perhaps get away with a simpler integration tests or a subset of their tests. Alternatively, we may also want to consider running these tests in parallel with something like https://github.com/grosser/parallel_tests.

This (in some form) will be a requirement for @anastasiasulyagina's GSoC project to be able to remove duplicated/vendored core code from Caskroom/cask and eventually move it into Homebrew itself and it's something we should have done for Homebrew/bundle a while ago. This also perhaps overlaps with @eirinikos's integration testing work.

CC @Homebrew/maintainers for thoughts as this is a fairly big change in how we do CI for this repository but it's something we'll have to do in some form.

xu-cheng commented on Library/Homebrew/cmd/tests.rb:
```diff
@@ -1,9 +1,34 @@
+
+    if ARGV.include? "--official-cmd-taps"
+      OFFICIAL_CMD_TAPS.each do |org, repo|
+        tap = "#{org}/#{repo}"
```
`tap = Tap.fetch(user, repo)`

xu-cheng commented on Library/Homebrew/cmd/tests.rb:
```diff
@@ -1,9 +1,34 @@
+    if ARGV.include? "--official-cmd-taps"
+      OFFICIAL_CMD_TAPS.each do |org, repo|
+        tap = "#{org}/#{repo}"
+        tap_dir = HOMEBREW_LIBRARY/"Taps/#{org}/homebrew-#{repo}"
```
You can use `tap.path`

xu-cheng commented on Library/Homebrew/cmd/tests.rb:
```diff
@@ -1,9 +1,34 @@
+        tap_dir = HOMEBREW_LIBRARY/"Taps/#{org}/homebrew-#{repo}"
+
+        # It has no tests :(
+        next if "#{tap}" == "homebrew/services"
```
`tap == "homebrew/services"`. We have implemented `Tap#==` for this kind of need.

UniqMartin replied:
But if `Tap#==` didn't exist, `tap.to_s` would still be preferable over `"#{tap}"`.

UniqMartin replied:
Should we maybe instead check for the presence of a `Rakefile` instead of hard-coding this? Though that would uselessly install the tap and then do nothing with it, if there is none.

xu-cheng commented on Library/Homebrew/cmd/tests.rb:
```diff
@@ -1,9 +1,34 @@
+        next if "#{tap}" == "homebrew/services"
+        ohai "Running #{tap} tests:"
+
+        system "brew", "tap", tap unless tap_dir.exist?
```
`system "brew", "tap", tap` You don't need to check tap existence.

MikeMcQuaid replied:
It warns otherwise.

xu-cheng replied:
If so, how about:
`system "brew", "tap", tap unless tap.installed?`
or
`tap.install unless tap.installed?`

UniqMartin replied:
If we stick with the `system` version, use `HOMEBREW_BREW_FILE` instead of `"brew"` like we do in most other places like [`Library/brew.rb`](https://github.com/Homebrew/brew/blob/915eed4c6427271c5b4c01fa36239361c3680c40/Library/brew.rb#L112-L113).

xu-cheng commented on Library/Homebrew/cmd/tests.rb:
```diff
@@ -1,9 +1,34 @@
+        system "brew", "tap", tap unless tap_dir.exist?
+        tap_dir.cd do
+          unless quiet_system("bundle", "check")
+            system "bundle", "install", "--path", "vendor/bundle"
```
Should we check `vendor/bundle` is in `.gitignore` for all theses taps.

MikeMcQuaid replied:
If we're not committing: I don't think it matters but probably not a bad idea anyway.

xu-cheng commented on Library/Homebrew/cmd/tests.rb:
```diff
@@ -1,9 +1,34 @@
+        end
+      end
+
+      return
```
So `--official-cmd-taps` won't run tests for homebrew?

MikeMcQuaid replied:
Maybe, yes.

UniqMartin replied:
Not sure if we need a separate command instead of using `tests` for that, but the whole `if ARGV.include? "--official-cmd-taps"` block should best live in a separate method. Otherwise the `tests` method becomes increasingly unreadable.

xu-cheng commented on Library/Homebrew/official_taps.rb:
```diff
@@ -15,4 +15,10 @@
+  ["caskroom", "cask"],
+  ["homebrew", "bundle"],
+  ["homebrew", "services"],
+].freeze
```
I think we can just store `homebrew/bundle`. Also if you want to support cmd, maybe:
```ruby
OFFICIAL_CMD_TAPS = {
  "homebrew/bundle" => %w[bundle, ...]
  ...
 }.freeze
```

So for `brew.rb`
```
possible_tap = OFFICIAL_CMD_TAPS.detect do |name, cmds|
  cmds.include? cmd
end.map { |name| Tap.fetch(name) }
```

MikeMcQuaid commented:
@UniqMartin @xu-cheng Neither of you are opposed to the overall concept of this PR?

UniqMartin commented on Library/Homebrew/cmd/tests.rb:
```diff
@@ -1,9 +1,34 @@
+            system "bundle", "install", "--path", "vendor/bundle"
+          end
+          system "bundle", "exec", "rake"
+          Homebrew.failed = !$?.success?
```
```ruby
Homebrew.failed = true unless $?.success?
```

Otherwise a later success may mask a previous failure.

UniqMartin commented on Library/Homebrew/cmd/tests.rb:
```diff
@@ -1,9 +1,34 @@
+    ENV["HOMEBREW_NO_ANALYTICS_THIS_RUN"] = "1"
+    Homebrew.install_gem_setup_path! "bundler"
+
+    if ARGV.include? "--official-cmd-taps"
```
Quickly getting into bike-shedding territory, but maybe `--official-external-commands`? Sounds a bit clearer to me.

UniqMartin commented:
> @UniqMartin @xu-cheng Neither of you are opposed to the overall concept of this PR?

I'm not opposed. Makes a lot of sense to me, though I personally don't care that much about these external commands because I don't use them. How long does this additional step approximately take to complete compared to a regular `brew tests` run? Does running this on our Mavericks test bot (with Ruby 1.8.7) make sense? Do those external commands even support Ruby 1.8.7?

xu-cheng commented:
I'm not oppose this PR's concept as well.

MikeMcQuaid commented:
Have addressed feedback.

> How long does this additional step approximately take to complete compared to a regular brew tests run? Does running this on our Mavericks test bot (with Ruby 1.8.7) make sense? Do those external commands even support Ruby 1.8.7?

Will get you the numbers for this after this PR runs the CI job.

UniqMartin commented on Library/brew.rb:
```diff
@@ -94,14 +94,8 @@ def require?(path)
-    when "services"
-      Tap.fetch("Homebrew", "services")
-    end
+    possible_tap = OFFICIAL_CMD_TAPS.find { |_, repo| repo == cmd }
```
Are you intentionally dropping support for auto-tapping when `brew bundle` is invoked with the alternative commands `brewdle`,  `brewdler`, and  `bundler`? (I'm not opposing this change, just making sure.)

xu-cheng replied:
Why don't we use the dictionary approach in the above? i.e.
```ruby
OFFICIAL_CMD_TAPS = {
  "homebrew/bundle" => %w[bundle, ...]
  ...
 }.freeze
```

I however opposite to only check cmd with tap name, which it's a bad assumption for future commands or taps.

MikeMcQuaid replied:
> Are you intentionally dropping support for auto-tapping when brew bundle is invoked with the alternative commands brewdle, brewdler, and bundler? (I'm not opposing this change, just making sure.)

Yes. PR incoming for homebrew-bundle to do that.



> I am however opposite to only check cmd with tap name, which it's a bad assumption for future commands or taps.

I don't think it's a bad assumption but we can correct it when it's no longer correct as it's an implementation detail.

xu-cheng replied:
If you take `homebrew/alias` into account, the assumption is already broken. 😉

UniqMartin replied:
This now either needs a `require "official_taps"` added before this line or directly in `tap.rb` (otherwise the constant `OFFICIAL_CMD_TAPS` is undefined at this point).

UniqMartin replied:
I also agree with @xu-cheng that it would be better to use his suggested hash approach from the start. Unlike the currently proposed solution it doesn't make any assumptions about the tap name vs. command name relationship and allows for multiple commands in a single tap, thus feels very future-proof to me. It doesn't feel like “doing it right” from the start creates a significant burden.

MikeMcQuaid replied:
The redundancy and speculation annoys me a bit here but will defer to you to get this merged.

UniqMartin commented on Library/Homebrew/dev-cmd/test-bot.rb:
```diff
@@ -659,6 +659,7 @@ def homebrew
         test "brew", "tests", *tests_args
         test "brew", "tests", "--no-compat"
         test "brew", "readall", "--syntax"
+        test "brew", "official-cmd-tap-tests"
```
Can you please add the file implementing this command to your commit? It's difficult to review and/or test otherwise. :wink:

MikeMcQuaid replied:
Yeh, whoops, re-added.

UniqMartin commented:
Have you been able to run `brew official-cmd-tap-tests` successfully? If I try that locally on my machine (OS X 10.11) in a Homebrew installation without any pre-installed taps except `homebew/core`, the Homebrew Cask tests actually fail with:

```console
$ brew official-cmd-tap-tests
==> Running caskroom/cask tests:
==> Tapping caskroom/cask
[…snip…]
8307 tests, 10105 assertions, 1 failures, 4 errors, 2395 skips
[…snip…]
```

The failure and those errors of course result in a nonzero exit code. On the other hand, the tests run for `homebrew/bundle` finish just fine.

----

Ignoring the failures for a moment, here are some numbers when running this locally:

```console
$ /usr/bin/time brew tests --no-compat
[…snip…]
       74.16 real        57.36 user        16.19 sys
$ /usr/bin/time brew tests --no-compat
[…snip…]
       67.33 real        54.76 user        13.66 sys
$ /usr/bin/time brew official-cmd-tap-tests
[…snip…]
      353.80 real       142.69 user        39.52 sys
$ /usr/bin/time brew official-cmd-tap-tests
[…snip…]
      193.10 real        76.73 user        30.89 sys
```

Those extra tests are not exactly cheap, particularly on the first run when the taps need to be installed and the Gems required for testing need to be fetched and installed via `bundle install`. On our test bot we'll obviously always have the slower times because all the taps and Gems will need to be installed anew on every test bot run.

MikeMcQuaid commented:
Cask's tests seem very oriented towards their Travis setup so I decided it was a bad idea to run the entire test suite (and probably overkill) so I decided to just write integration tests to smoke tests these three commands; in my experience if we e.g. break an API they are relying on we'll realise pretty quickly. I think this gives us the best of both worlds; fast integration tests (less than a second added in total for all three assuming they are already tapped) whilst making sure we don't obliterate the internal methods they rely on.

UniqMartin commented on Library/Homebrew/test/test_integration_cmds.rb:
```diff
@@ -668,4 +674,32 @@ def test_search
+
+  def test_services
+    skip_unless_test_cmd_taps
+    assert_equal "Warning: No services available to control with `brew services`", cmd("services", "list")
```
Can you wrap this overly long line, just like you did in the `bundle check` case?

UniqMartin commented on Library/Homebrew/test/test_integration_cmds.rb:
```diff
@@ -668,4 +674,32 @@ def test_search
+        cmd("bundle", "check")
+    end
+  ensure
+    (HOMEBREW_REPOSITORY/".git").rmtree
```
This is causing trouble, as it is also executed if you skip the test due to the `skip_unless_test_cmd_taps` at the top of this test method.

MikeMcQuaid replied:
@UniqMartin 👍. I think we should probably be using `FileUtils.rm_rf` for all these cases to avoid the failure when it doesn't exist and avoid guarding them all with `.exist?`

UniqMartin replied:
Probably not for this PR, but we could also consider adding a method that basically does “register this path for cleanup” and that is called every time a new path in a test is used. This would delegate the removal of files and directories to `teardown` instead of relying on `ensure` in every single test method (where this file/directory cleanup is dominating). I'm not sure though, whether this is the right approach or if it makes the tests less readable.

UniqMartin commented on Library/Homebrew/test/test_integration_cmds.rb:
```diff
@@ -17,6 +17,12 @@ def teardown
     @formula_files.each(&:unlink)
   end
 
+  def skip_unless_test_cmd_taps
```
We have a very similar method in `test/testing_env.rb` named `needs_compat`. Would you consider renaming this one to `needs_test_cmd_taps` for consistency? (I don't disagree that the currently used method name is appropriate if viewed in isolation.)

UniqMartin commented on Library/Homebrew/cmd/tests.rb:
```diff
@@ -3,11 +3,21 @@
   def tests
+    ENV["HOMEBREW_NO_ANALYTICS_THIS_RUN"] = "1"
+
+    if ARGV.include? "--official-cmd-taps"
```
Would you mind adding this option to the documentation comment at the top of the file?

UniqMartin commented:
Commented some code, that is probably causing the test bot run to fail, and on a few details.

> Cask's tests seem very oriented towards their Travis setup so I decided it was a bad idea to run the entire test suite (and probably overkill)

I agree.

> so I decided to just write integration tests to smoke tests these three commands; in my experience if we e.g. break an API they are relying on we'll realise pretty quickly.

Makes sense and works for me.

UniqMartin commented on Library/Homebrew/cmd/tests.rb:
```diff
@@ -1,13 +1,22 @@
+    if ARGV.include? "--official-cmd-taps"
+      ENV["HOMEBREW_TEST_OFFICIAL_CMD_TAPS"] = "1"
+      OFFICIAL_CMD_TAPS.each do |tap, _|
+        tap = Tap.fetch tap
```
This probably needs a `require "tap"` at the top of the file. Here's what I currently get:

```console
$ brew tests --no-compat --official-cmd-taps
Error: uninitialized constant Homebrew::Tap
Please report this bug:
    https://git.io/brew-troubleshooting
/opt/brewery/dummy/Library/Homebrew/cmd/tests.rb:14:in `block in tests'
/opt/brewery/dummy/Library/Homebrew/cmd/tests.rb:13:in `each'
/opt/brewery/dummy/Library/Homebrew/cmd/tests.rb:13:in `tests'
/opt/brewery/dummy/Library/brew.rb:87:in `<main>'
```

UniqMartin commented:
I'm sorry, but I had to revert this in dba1958bd79c1c9d18f215dfc2b806ea62edd1c8. This needs more work and local testing via, e.g.,

```
brew tests --official-cmd-taps
```

before it can be merged (as the test bot won't run this code until it is merged, at which point it's too late). To quote from the revert commit:

> Taps installed prior to running the test suite are not visible to the test suite as most Homebrew paths are redefined as to not mess up the local installation.

UniqMartin commented:
I just realized that there is/was another issue with the current implementation. The entire test suite now runs thrice: once with coverage in no-compat mode, once regularly, and now once more with the three integration tests from this PR enabled. The latter two runs are basically identical except for those newly added integration tests, resulting in some unnecessary and wasteful redundancy.

MikeMcQuaid commented:
Addressed this in a series of commits ending in https://github.com/Homebrew/brew/commit/a1b0ef1300ff8ac07ef904479b0321a2a5fc0368 as currently the bot isn't doing CI testing with the correct version so it was easier to just ensure it was working properly on all platforms by pushing to `master`. The green builds for Travis and Jenkins are: https://travis-ci.org/Homebrew/brew/builds/142266951 http://bot.brew.sh/job/Homebrew%20Testing/1029/
